rclpy/src/rclpy/_rclpy.c

```diff
@@ -6068,98 +6033,94 @@ static PyMethodDef rclpy_methods[] = {
-    Py_DECREF(m);
-    return NULL;
+
+  RCLError = PyObject_GetAttrString(exceptions_module, "RCLError");
+  if (NULL == RCLError) {
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  RCLInvalidROSArgsError = PyErr_NewExceptionWithDoc(
-    "_rclpy.RCLInvalidROSArgsError",
-    "Thrown when invalid ROS arguments are found by rcl.",
-    RCLError, NULL);
+  RCLInvalidROSArgsError = PyObject_GetAttrString(exceptions_module, "RCLInvalidROSArgsError");
   if (NULL == RCLInvalidROSArgsError) {
-    Py_DECREF(m);
-    return NULL;
-  }
-  if (PyModule_AddObject(m, "RCLInvalidROSArgsError", RCLInvalidROSArgsError) != 0) {
-    Py_DECREF(m);
-    return NULL;
+    Py_DECREF(RCLError);
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  UnknownROSArgsError = PyErr_NewExceptionWithDoc(
-    "_rclpy.UnknownROSArgsError",
-    "Thrown when unknown ROS arguments are found.",
-    RCLError, NULL);
+  UnknownROSArgsError = PyObject_GetAttrString(exceptions_module, "UnknownROSArgsError");
   if (NULL == UnknownROSArgsError) {
-    Py_DECREF(m);
-    return NULL;
-  }
-  if (PyModule_AddObject(m, "UnknownROSArgsError", UnknownROSArgsError) != 0) {
-    Py_DECREF(m);
-    return NULL;
+    Py_DECREF(RCLInvalidROSArgsError);
+    Py_DECREF(RCLError);
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  NodeNameNonExistentError = PyErr_NewExceptionWithDoc(
-    "_rclpy.NodeNameNonExistentError",
-    "Thrown when a node name is not found.",
-    RCLError, NULL);
-  if (PyModule_AddObject(m, "NodeNameNonExistentError", NodeNameNonExistentError)) {
-    Py_DECREF(m);
-    return NULL;
+  NodeNameNonExistentError = PyObject_GetAttrString(exceptions_module, "NodeNameNonExistentError");
+  if (NULL == NodeNameNonExistentError) {
+    Py_DECREF(UnknownROSArgsError);
+    Py_DECREF(RCLInvalidROSArgsError);
+    Py_DECREF(RCLError);
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  UnsupportedEventTypeError = PyErr_NewExceptionWithDoc(
-    "_rclpy.UnsupportedEventTypeError",
-    "Thrown when registering a callback for an event type that is not supported.",
-    RCLError, NULL);
-  if (PyModule_AddObject(m, "UnsupportedEventTypeError", UnsupportedEventTypeError)) {
-    Py_DECREF(m);
-    return NULL;
+  UnsupportedEventTypeError =
+    PyObject_GetAttrString(exceptions_module, "UnsupportedEventTypeError");
+  if (NULL == UnsupportedEventTypeError) {
+    Py_DECREF(NodeNameNonExistentError);
+    Py_DECREF(UnknownROSArgsError);
+    Py_DECREF(RCLInvalidROSArgsError);
+    Py_DECREF(RCLError);
+    Py_DECREF(exceptions_module);
+    return -1;
   }
 
-  if (PyErr_Occurred()) {
-    Py_DECREF(m);
-    return NULL;
-  }
-  return m;
+  Py_DECREF(exceptions_module);
+
+  // 0 is success
+  return 0;
+}
+
+static PyModuleDef_Slot rclpy_slots[] = {
+  {Py_mod_exec, rclpy_exec},
+  {0, NULL}
+};
+
+/// Define the Python module
+static struct PyModuleDef _rclpymodule = {
+  PyModuleDef_HEAD_INIT,
+  "_rclpy",
+  rclpy__doc__,
+  0,
```

Comment:
Could you elaborate on why this changed from `-1` to `0`?

Reply:
It's the size of the [module's global state](https://docs.python.org/3/c-api/module.html#c.PyModuleDef.m_size). The reason for the change to `0` is `Non-negative m_size is required for multi-phase initialization`, and multi-phase initialization is needed to allow this module to import the new exceptions module.
 
The docs say -1 means there is state in global variables, and so it would be unsafe to use multiple sub-interpretters. Looking at this again, there are global variables storing references to the exceptions. IIUC this change means another change is needed to store those in the module's state instead.
https://github.com/ros2/rclpy/blob/812ea5082d008dc5c94a7d49e69df2d55415eec5/rclpy/src/rclpy/_rclpy.c#L53-L57

Reply:
Great explanation, thanks. Makes sense to me.

> ...another change is needed to store those in the module's state instead.

Are you planning to amend this PR or follow-up with another?

Reply:
Amend this one

Reply:
Uses module state instead of global variables in 30621b6